### PR TITLE
ECO-2637: added support of cash on delivery

### PR DESCRIPTION
**Pull request by Ruslan05**
- Developer(s): @Ruslan05

- Ticket: https://spryker.atlassian.net/browse/ECO-2637

#### Please confirm

- [x] No new OS components - or they have been approved by legal department

#### Documentation

- [x] Functional documentation provided or in progress?
- [x] Integration guide for projects provided or not needed?
- [x] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Payone                | minor                 |                       |

#### Release Notes

{skip}

-----------------------------------------

##### Change log

Improvements

- Added new property to `PaymentDetail ` and `Payment ` transfers.
- Added new transfer `PayonePaymentCashOnDelivery `.
- Added new plugin `PayoneCashOnDeliverySubFormPlugin`
- Added new api constant `PAYMENT_METHOD_CASH_ON_DELIVERY_OMS `.

### CODE REVIEW

File: src/SprykerEco/Yves/Payone/Form/DataProvider/CashOnDeliveryDataProvider.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+use Spryker\Yves\StepEngine\Dependency\Form\StepEngineFormDataProviderInterface;
+
+class CashOnDeliveryDataProvider implements StepEngineFormDataProviderInterface
+{
+    /**
+     * @param \Spryker\Shared\Kernel\Transfer\AbstractTransfer|\Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     *
+     * @return \Spryker\Shared\Kernel\Transfer\AbstractTransfer
+     */
+    public function getData(AbstractTransfer $quoteTransfer)
```
Comment (vhrychenko) [inline src/SprykerEco/Yves/Payone/Form/DataProvider/CashOnDeliveryDataProvider.php]: Return type

File: src/SprykerEco/Yves/Payone/Form/DataProvider/CashOnDeliveryDataProvider.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        }
+        return $quoteTransfer;
+    }
+
+    /**
+     * @param \Spryker\Shared\Kernel\Transfer\AbstractTransfer $quoteTransfer
+     *
+     * @return array
+     */
+    public function getOptions(AbstractTransfer $quoteTransfer)
```
Comment (vhrychenko) [inline src/SprykerEco/Yves/Payone/Form/DataProvider/CashOnDeliveryDataProvider.php]: Return type

File: src/SprykerEco/Yves/Payone/Plugin/PayoneCashOnDeliverySubFormPlugin.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * @method \SprykerEco\Yves\Payone\PayoneFactory getFactory()
+ */
+class PayoneCashOnDeliverySubFormPlugin extends AbstractPlugin implements SubFormPluginInterface
+{
+    /**
+     * @return \SprykerEco\Yves\Payone\Form\CashOnDeliverySubForm
+     */
+    public function createSubForm()
```
Comment (vhrychenko) [inline src/SprykerEco/Yves/Payone/Plugin/PayoneCashOnDeliverySubFormPlugin.php]: Please check return types everywhere

**Comment by Ruslan05** [comment]
I was following it`s interface, to have same method`s signature:
https://github.com/spryker/spryker/blob/master/Bundles/StepEngine/src/Spryker/Yves/StepEngine/Dependency/Form/StepEngineFormDataProviderInterface.php

Should i still add return types ?

File: src/SprykerEco/Yves/Payone/Form/CashOnDeliverySubForm.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+use Generated\Shared\Transfer\PayonePaymentTransfer;
+use Spryker\Yves\StepEngine\Dependency\Form\SubFormInterface;
+use SprykerEco\Shared\Payone\PayoneConstants;
+use Symfony\Component\Form\Extension\Core\Type\HiddenType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class CashOnDeliverySubForm extends AbstractPayoneSubForm
+{
+    const PAYMENT_METHOD = 'cash_on_delivery';
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Yves/Payone/Form/CashOnDeliverySubForm.php]: Please add visibility for this constant.

File: src/SprykerEco/Yves/Payone/Form/DataProvider/CashOnDeliveryDataProvider.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+use Generated\Shared\Transfer\PaymentTransfer;
+use Generated\Shared\Transfer\PayonePaymentCashOnDeliveryTransfer;
+use Generated\Shared\Transfer\PayonePaymentTransfer;
+use Spryker\Shared\Kernel\Transfer\AbstractTransfer;
+use Spryker\Yves\StepEngine\Dependency\Form\StepEngineFormDataProviderInterface;
+
+class CashOnDeliveryDataProvider implements StepEngineFormDataProviderInterface
+{
+    /**
+     * @param \Spryker\Shared\Kernel\Transfer\AbstractTransfer|\Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Yves/Payone/Form/DataProvider/CashOnDeliveryDataProvider.php]: Change signature to QuoteTransfer only for parameter and return type.

File: src/SprykerEco/Yves/Payone/Form/DataProvider/CashOnDeliveryDataProvider.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        if ($quoteTransfer->getPayment() === null) {
+            $paymentTransfer = new PaymentTransfer();
+            $paymentTransfer->setPayone(new PayonePaymentTransfer());
+            $quoteTransfer->setPayment($paymentTransfer);
+        }
+        return $quoteTransfer;
+    }
+
+    /**
+     * @param \Spryker\Shared\Kernel\Transfer\AbstractTransfer $quoteTransfer
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Yves/Payone/Form/DataProvider/CashOnDeliveryDataProvider.php]: Change signature to QuoteTransfer.

**Comment by vhrychenko** [comment]
As far as I know, you are able to add a return type to child classes without method signature errors.  @Ruslan05

File: src/SprykerEco/Yves/Payone/Handler/PayoneHandler.php [hunk 7]
```diff
@@ -157,6 +159,12 @@ protected function setPayonePayment(Request $request, QuoteTransfer $quoteTransf
             $paymentDetailTransfer->setBankGroupType($payonePaymentTransfer->getBankGroupType());
             $paymentDetailTransfer->setIban($payonePaymentTransfer->getIban());
             $paymentDetailTransfer->setBic($payonePaymentTransfer->getBic());
+        } elseif ($paymentSelection == PaymentTransfer::PAYONE_CASH_ON_DELIVERY) {
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Yves/Payone/Handler/PayoneHandler.php]: This is worst place of the module... It must be refactored, but maybe not in this PR.

File: src/SprykerEco/Yves/Payone/Handler/PayoneHandler.php [hunk 8]
```diff
@@ -157,6 +159,12 @@ protected function setPayonePayment(Request $request, QuoteTransfer $quoteTransf
...
+            $shippingProvider = $quoteTransfer->getShipment()->getMethod()->getCarrierName();
+            $payonePaymentTransfer = $this->getPayonePaymentTransfer($quoteTransfer, $paymentSelection);
+
+            $payonePaymentTransfer->setShippingProvider($shippingProvider);
+            $paymentDetailTransfer->setShippingProvider($shippingProvider);
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Yves/Payone/Handler/PayoneHandler.php]: Why do you need this twice?

File: src/SprykerEco/Yves/Payone/PayoneFactory.php [hunk 9]
```diff
@@ -89,6 +91,22 @@ public function createInvoiceSubFormDataProvider()
         return new InvoiceDataProvider();
     }
 
+    /**
+     * @return \SprykerEco\Yves\Payone\Form\CashOnDeliverySubForm
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Yves/Payone/PayoneFactory.php]: Change signature to interface.

File: src/SprykerEco/Yves/Payone/PayoneFactory.php [hunk 10]
```diff
@@ -89,6 +91,22 @@ public function createInvoiceSubFormDataProvider()
...
+     */
+    public function createCashOnDeliverySubForm(): SubFormInterface
+    {
+        return new CashOnDeliverySubForm();
+    }
+
+    /**
+     * @return \SprykerEco\Yves\Payone\Form\DataProvider\CashOnDeliveryDataProvider
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Yves/Payone/PayoneFactory.php]: Change signature to interface.

File: src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+use SprykerEco\Zed\Payone\Business\Api\Request\Container\PreAuthorizationContainer;
+use SprykerEco\Zed\Payone\Business\Api\Request\Container\RefundContainer;
+use SprykerEco\Zed\Payone\Dependency\Facade\PayoneToGlossaryInterface;
+
+class CashOnDelivery extends AbstractMapper
+{
+    /**
+     * @var \SprykerEco\Zed\Payone\Dependency\Facade\PayoneToGlossaryInterface
+     */
+    protected $glossary;
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php]: Rename it `$glossaryFacade`.
I don’t if it was you who added bridge but it has to be renamed to `PayoneToGlossaryFacadeInterface` and `PayoneToGlossaryFacadeBridge`.

File: src/SprykerEco/Zed/Payone/Business/PayoneBusinessFactory.php [hunk 12]
```diff
@@ -304,6 +307,16 @@ protected function createEWallet($storeConfig)
         return $EWallet;
     }
 
+    /**
+     * @param $storeConfig
+     *
+     * @return \SprykerEco\Zed\Payone\Business\Payment\MethodMapper\CashOnDelivery
+     */
+    protected function createCashOnDelivery(Store $storeConfig, PayoneToGlossaryInterface $glossary): CashOnDelivery
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Zed/Payone/Business/PayoneBusinessFactory.php]: Please change return type to interface if it's present for this type of models.

File: src/SprykerEco/Yves/Payone/Handler/PayoneHandler.php (hunk 7 above)
Comment (Ruslan05) [reply src/SprykerEco/Yves/Payone/Handler/PayoneHandler.php]: Agree, but such a refactoring is out of the scope

File: src/SprykerEco/Yves/Payone/Form/DataProvider/CashOnDeliveryDataProvider.php [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+     * @return \Spryker\Shared\Kernel\Transfer\AbstractTransfer
+     */
+    public function getData(AbstractTransfer $quoteTransfer)
+    {
+        if ($quoteTransfer->getPayment() === null) {
+            $paymentTransfer = new PaymentTransfer();
+            $paymentTransfer->setPayone(new PayonePaymentTransfer());
+            $quoteTransfer->setPayment($paymentTransfer);
+        }
+        return $quoteTransfer;
```
Comment (gechetspr) [inline src/SprykerEco/Yves/Payone/Form/DataProvider/CashOnDeliveryDataProvider.php]: add an empty line before return statement

File: src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+use SprykerEco\Zed\Payone\Business\Api\Request\Container\Authorization\PersonalContainer;
+use SprykerEco\Zed\Payone\Business\Api\Request\Container\AuthorizationContainer;
+use SprykerEco\Zed\Payone\Business\Api\Request\Container\BankAccountCheckContainer;
+use SprykerEco\Zed\Payone\Business\Api\Request\Container\CaptureContainer;
+use SprykerEco\Zed\Payone\Business\Api\Request\Container\DebitContainer;
+use SprykerEco\Zed\Payone\Business\Api\Request\Container\PreAuthorizationContainer;
+use SprykerEco\Zed\Payone\Business\Api\Request\Container\RefundContainer;
+use SprykerEco\Zed\Payone\Dependency\Facade\PayoneToGlossaryInterface;
+
+class CashOnDelivery extends AbstractMapper
```
Comment (gechetspr) [inline src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php]: Shouldn't it be smth like `CashOnDeliveryMapper`?

File: src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+        return PayoneApiConstants::PAYMENT_METHOD_CASH_ON_DELIVERY;
+    }
+
+    /**
+     * @param \Orm\Zed\Payone\Persistence\SpyPaymentPayone $paymentEntity
+     * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
+     *
+     * @return \SprykerEco\Zed\Payone\Business\Api\Request\Container\Authorization\AbstractAuthorizationContainer|\SprykerEco\Zed\Payone\Business\Api\Request\Container\AuthorizationContainer
+     */
+    public function mapPaymentToAuthorization(SpyPaymentPayone $paymentEntity, OrderTransfer $orderTransfer): AbstractAuthorizationContainer
```
Comment (gechetspr) [inline src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php]: Why do we have an Entity in the business layer?

File: src/SprykerEco/Yves/Payone/Handler/PayoneHandler.php (hunk 8 above)
Comment (Ruslan05) [reply src/SprykerEco/Yves/Payone/Handler/PayoneHandler.php]: Removed unnecessary one `$payonePaymentTransfer->setShippingProvider($shippingProvider);`

File: src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+
+        $authorizationContainer->setAid($this->getStandardParameter()->getAid());
+        $authorizationContainer->setClearingType(PayoneApiConstants::CLEARING_TYPE_CASH_ON_DELIVERY);
+        $authorizationContainer->setReference($paymentEntity->getReference());
+        $authorizationContainer->setAmount($paymentDetailEntity->getAmount());
+        $authorizationContainer->setCurrency($this->getStandardParameter()->getCurrency());
+        $authorizationContainer->setPaymentMethod($this->createPaymentMethodContainerFromPayment($paymentEntity));
+
+        $personalContainer = new PersonalContainer();
+        $this->mapBillingAddressToPersonalContainer($personalContainer, $paymentEntity);
```
Comment (gechetspr) [inline src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php]: mapBillingAddressToPersonalContainer returns void?
In any case, it looks that you can move this to the separate method

File: src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+    /**
+     * @param \Orm\Zed\Payone\Persistence\SpyPaymentPayone $paymentEntity
+     *
+     * @return \SprykerEco\Zed\Payone\Business\Api\Request\Container\Authorization\PaymentMethod\CashOnDeliveryContainer
+     */
+    protected function createPaymentMethodContainerFromPayment(SpyPaymentPayone $paymentEntity): CashOnDeliveryContainer
+    {
+        $shippingProviderName = $paymentEntity->getSpyPaymentPayoneDetail()->getShippingProvider();
+
+        $translatedShippingProviderName = $shippingProviderName ? $this->translate($shippingProviderName) : $shippingProviderName;
```
Comment (gechetspr) [inline src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php]: So we might have no shipping provider? Are you sure that in this case we should just continue?

File: src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php (hunk 11 above)
Comment (Ruslan05) [reply src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php]: This is an old one, renamed

File: src/SprykerEco/Zed/Payone/Dependency/Facade/PayoneToGlossaryBridge.php [hunk 18]
```diff
@@ -36,6 +36,19 @@ public function translateByKeyId($idKey, array $data = [])
         return $this->glossaryFacade->translateByKeyId($idKey, $data);
     }
 
+    /**
+     * @param $keyName
+     * @param array $data
+     * @param \Generated\Shared\Transfer\LocaleTransfer|null $localeTransfer
+     *
+     * @return string
+     * @throws \Spryker\Zed\Glossary\Business\Exception\MissingTranslationException
```
Comment (gechetspr) [inline src/SprykerEco/Zed/Payone/Dependency/Facade/PayoneToGlossaryBridge.php]: You don't need this

File: tests/SprykerEcoTest/Zed/Payone/Business/Payment/MethodMapper/CashOnDeliveryTest.php [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+ * @group Zed
+ * @group Payone
+ * @group Business
+ * @group Payment
+ * @group MethodMapper
+ * @group CashOnDeliveryTest
+ */
+class CashOnDeliveryTest extends AbstractMethodMapperTest
+{
+    const STANDARD_PARAMETER_CLEARING_TYPE = 'cod';
```
Comment (gechetspr) [inline tests/SprykerEcoTest/Zed/Payone/Business/Payment/MethodMapper/CashOnDeliveryTest.php]: Please add visibility

File: tests/SprykerEcoTest/Zed/Payone/Business/Payment/MethodMapper/CashOnDeliveryTest.php [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+        'clearingtype' => self::STANDARD_PARAMETER_CLEARING_TYPE,
+        'reference' => self::PAYMENT_REFERENCE,
+        'amount' => self::AMOUNT_FULL,
+        'currency' => self::STANDARD_PARAMETER_CURRENCY,
+    ];
+
+    /**
+     * @return void
+     */
+    public function testMapPaymentToPreauthorization()
```
Comment (gechetspr) [inline tests/SprykerEcoTest/Zed/Payone/Business/Payment/MethodMapper/CashOnDeliveryTest.php]: and return types

File: tests/SprykerEcoTest/Zed/Payone/Business/Payment/MethodMapper/CashOnDeliveryTest.php [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+        foreach (static::DEBIT_COMMON_REQUIRED_PARAMS as $key => $value) {
+            $this->assertArrayHasKey($key, $requestData);
+            $this->assertSame($value, $requestData[$key]);
+        }
+    }
+
+    /**
+     * @return \Orm\Zed\Payone\Persistence\SpyPaymentPayoneDetail
+     */
+    protected function getPaymentPayoneDetailMock()
```
Comment (gechetspr) [inline tests/SprykerEcoTest/Zed/Payone/Business/Payment/MethodMapper/CashOnDeliveryTest.php]: You don't need this method

File: tests/SprykerEcoTest/Zed/Payone/_support/_generated/PayoneZedTesterActions.php [hunk 22]
```diff
@@ -1,4 +1,4 @@
-<?php  //[STAMP] 6d265cc3529a8d3f945309a25b014809
```
Comment (gechetspr) [inline tests/SprykerEcoTest/Zed/Payone/_support/_generated/PayoneZedTesterActions.php]: Shouldn't this be ignored by git?

File: src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php (hunk 14 above)
Comment (Ruslan05) [reply src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php]: Please check other similar classes, they named accordingly, i just extenteded it with provided project files, in module level

File: src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php (hunk 15 above)
Comment (Ruslan05) [reply src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php]: Same as above

File: src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php (hunk 16 above)
Comment (Ruslan05) [reply src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php]: It`s very old code...moved partially to `buildPersonalContainer()`

File: tests/SprykerEcoTest/Zed/Payone/Business/Payment/MethodMapper/CashOnDeliveryTest.php (hunk 21 above)
Comment (Ruslan05) [reply tests/SprykerEcoTest/Zed/Payone/Business/Payment/MethodMapper/CashOnDeliveryTest.php]: removed

File: tests/SprykerEcoTest/Zed/Payone/_support/_generated/PayoneZedTesterActions.php (hunk 22 above)
Comment (Ruslan05) [reply tests/SprykerEcoTest/Zed/Payone/_support/_generated/PayoneZedTesterActions.php]: have no idea, i can add to .gitignore if you request ?

File: src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php (hunk 17 above)
Comment (Ruslan05) [reply src/SprykerEco/Zed/Payone/Business/Payment/MethodMapper/CashOnDelivery.php]: We will receive Flash message from API in this case, which says `invalid shipment provider`

File: src/SprykerEco/Zed/Payone/Business/Internal/Installer.php [hunk 23]
```diff
@@ -26,10 +26,10 @@ class Installer extends AbstractInstaller
     protected $config;
 
     /**
-     * @param \SprykerEco\Zed\Payone\Dependency\Facade\PayoneToGlossaryInterface $glossary
+     * @param \SprykerEco\Zed\Payone\Dependency\Facade\PayoneToGlossaryFacadeInterface $glossary
      * @param \SprykerEco\Zed\Payone\PayoneConfig $config
      */
-    public function __construct(PayoneToGlossaryInterface $glossary, PayoneConfig $config)
+    public function __construct(PayoneToGlossaryFacadeInterface $glossary, PayoneConfig $config)
```
Comment (gechetspr) [inline src/SprykerEco/Zed/Payone/Business/Internal/Installer.php]: It would be great if you can rename it to glossaryFacade too

**Pull request merged**